File: shared/kernel.js
// shared/kernel.js
// The CostSauce costing kernel, JavaScript implementation. ES module, zero
// dependencies, Node >= 18 or any modern browser.
//
// Contract: docs/superpowers/specs/2026-07-25-native-ios-app-design.md
// §8-§10, pinned by shared/golden-vectors.json (also run by the Python
// kernel). All decimals cross this API as STRINGS; internals are BigInt
// rationals. Rounding is half-away-from-zero everywhere.

export class KernelError extends Error {}

// ---- rational plumbing ----------------------------------------------------
function parseDec(s) {
  const m = /^(-?)(\d+)(?:\.(\d+))?$/.exec(String(s).trim());
  if (!m) throw new KernelError(`not a decimal: ${s}`);
  const [, sign, ip, fp = ""] = m;
  const n = BigInt(ip + fp) * (sign === "-" ? -1n : 1n);
  return { n, d: 10n ** BigInt(fp.length) };
}
const ratMul = (a, b) => ({ n: a.n * b.n, d: a.d * b.d });
function ratDiv(a, b) {
  if (b.n === 0n) throw new KernelError("division by zero");
  let n = a.n * b.d, d = a.d * b.n;
  if (d < 0n) { n = -n; d = -d; }        // keep denominators positive
  return { n, d };
}
const ratAdd = (a, b) => ({ n: a.n * b.d + b.n * a.d, d: a.d * b.d });
const ratSub = (a, b) => ({ n: a.n * b.d - b.n * a.d, d: a.d * b.d });
function ratCmp(a, b) {
  const x = a.n * b.d - b.n * a.d;
  return x < 0n ? -1 : x > 0n ? 1 : 0;
}
const ratPos = (a) => a.n > 0n;

export function roundHalfAway(rat, places) {
  const scale = 10n ** BigInt(places);
  let { n, d } = rat;
  const neg = n < 0n;
  if (neg) n = -n;
  const q = (2n * n * scale + d) / (2n * d);   // floor(x*scale + 0.5)
  const sign = neg && q > 0n ? "-" : "";
  if (places === 0) return sign + q.toString();
  const ip = q / scale, fp = q % scale;
  return `${sign}${ip}.${fp.toString().padStart(places, "0")}`;
}

// ---- names ----------------------------------------------------------------
export function normalizeName(name) {
  let s = String(name).toLowerCase().trim();
  s = s.replace(/[^a-z0-9\s]/g, "");
  s = s.replace(/\s+/g, " ").trim();
  if (s.endsWith("s") && !s.endsWith("ss") && s.length > 3) s = s.slice(0, -1);
  return s;
}

export function matchIngredient(name, candidates) {
  const norm = normalizeName(name);
  if (!norm) return null;
  for (const [id, cname] of candidates)
    if (normalizeName(cname) === norm) return [id, cname, "exact"];
  for (const [id, cname] of candidates) {
    const cn = normalizeName(cname);
    if (cn.includes(norm) || norm.includes(cn)) return [id, cname, "fuzzy"];
  }
  return null;
}

// ---- units & money --------------------------------------------------------
const WEIGHT_TO_LB = {
  lb: parseDec("1"), oz: { n: 1n, d: 16n },
  kg: parseDec("2.2046226218"), g: parseDec("0.0022046226218"),
};
const BASE_UNITS = ["lb", "oz", "kg", "g", "each"];

export function normalizePurchase({ baseUnit, qty, unit, totalPrice, qtyInCase }) {
  const q = parseDec(qty), t = parseDec(totalPrice);
  if (!ratPos(q) || !ratPos(t))
    throw new KernelError("qty and total_price must be positive");
  if (!BASE_UNITS.includes(baseUnit))
    throw new KernelError(`unknown base_unit ${baseUnit}`);
  unit = String(unit || "").trim().toLowerCase();
  let baseQty;
  if (unit === "case") {
    if (qtyInCase == null) throw new KernelError("qty_in_case required");
    const qc = parseDec(qtyInCase);
    if (!ratPos(qc)) throw new KernelError("qty_in_case must be positive");
    baseQty = ratMul(q, qc);
  } else if (baseUnit === "each") {
    if (unit !== "each")
      throw new KernelError("tracked 'each' — use unit 'each' or 'case'");
    baseQty = q;
  } else {
    if (!(unit in WEIGHT_TO_LB))
      throw new KernelError(`unsupported unit ${unit}`);
    baseQty = ratDiv(ratMul(q, WEIGHT_TO_LB[unit]), WEIGHT_TO_LB[baseUnit]);
  }
  const result = roundHalfAway(baseQty, 4);
  if (parseDec(result).n <= 0n)
    throw new KernelError(
      "quantity is too small to register at 4 decimal places");
  return result;
}

export function unitPrice(totalPrice, qtyBaseUnits) {
  const t = parseDec(totalPrice), q = parseDec(qtyBaseUnits);
  if (!ratPos(t) || !ratPos(q))
    throw new KernelError("total_price and qty_base_units must be positive");
  return roundHalfAway(ratDiv(t, q), 6);
}

export function suggestedPriceCents(plateCents, targetBp) {
  if (targetBp <= 0) throw new KernelError("target_bp must be positive");
  if (plateCents < 0) throw new KernelError("plate_cents must be non-negative");
  const num = BigInt(plateCents) * 10000n, den = BigInt(targetBp) * 50n;
  const ceil = (num + den - 1n) / den;
  return Number(ceil * 50n);
}

export function fcStatus(plateCents, menuCents, targetBp) {
  if (menuCents <= 0) throw new KernelError("menu_cents must be positive");
  const fc = roundHalfAway({ n: BigInt(plateCents) * 100n, d: BigInt(menuCents) }, 1);
  const fcR = parseDec(fc), target = { n: BigInt(targetBp), d: 100n };
  let status;
  if (ratCmp(fcR, target) <= 0) status = "ok";
  else if (ratCmp(fcR, ratAdd(target, { n: 2n, d: 1n })) <= 0) status = "watch";
  else status = "danger";
  return { fc, status };
}

// ---- drift ----------------------------------------------------------------
function dayNumber(iso) {
  // days since epoch via integer civil-date math — never Date (spec §4.3;
  // B5 was a Date-object timezone bug).
  const [y, m, d] = iso.split("-").map(Number);
  const yy = m <= 2 ? y - 1 : y;
  const era = Math.floor(yy / 400);
  const yoe = yy - era * 400;
  const doy = Math.floor((153 * (m + (m > 2 ? -3 : 9)) + 2) / 5) + d - 1;
  const doe = yoe * 365 + Math.floor(yoe / 4) - Math.floor(yoe / 100) + doy;
  return era * 146097 + doe - 719468;
}
function dayToIso(z) {
  z += 719468;
  const era = Math.floor(z / 146097);
  const doe = z - era * 146097;
  const yoe = Math.floor(
    (doe - Math.floor(doe / 1460) + Math.floor(doe / 36524)
     - Math.floor(doe / 146096)) / 365);
  const y = yoe + era * 400;
  const doy = doe - (365 * yoe + Math.floor(yoe / 4) - Math.floor(yoe / 100));
  const mp = Math.floor((5 * doy + 2) / 153);
  const d = doy - Math.floor((153 * mp + 2) / 5) + 1;
  const m = mp + (mp < 10 ? 3 : -9);
  const yy = m <= 2 ? y + 1 : y;
  return `${String(yy).padStart(4, "0")}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}

export function drift(rows) {
  const live = rows.filter((r) => !r.deleted);
  if (live.length === 0) return null;
  const ordered = [...live].sort((a, b) => {
    if (a.purchased_on !== b.purchased_on)
      return a.purchased_on < b.purchased_on ? 1 : -1;
    if (a.recorded_at !== b.recorded_at)
      return a.recorded_at < b.recorded_at ? 1 : -1;
    return a.id < b.id ? 1 : a.id > b.id ? -1 : 0;
  });
  const latest = ordered[0];
  const windowStartDay = dayNumber(latest.purchased_on) - 90;
  const windowStart = dayToIso(windowStartDay);
  const baseline = ordered.slice(1)
    .filter((r) => dayNumber(r.purchased_on) >= windowStartDay);
  const n = baseline.length;
  const out = { latest_price: latest.unit_price,
                latest_on: latest.purchased_on, window_start: windowStart,
                baseline_n: n, trailing_avg: null, drift_pct: null };
  if (n === 0) return out;
  let sum = { n: 0n, d: 1n };
  for (const r of baseline) sum = ratAdd(sum, parseDec(r.unit_price));
  const avg = ratDiv(sum, { n: BigInt(n), d: 1n });
  out.trailing_avg = roundHalfAway(avg, 6);
  if (n >= 3) {
    const pct = ratMul(ratDiv(ratSub(parseDec(latest.unit_price), avg), avg),
                       { n: 100n, d: 1n });
    out.drift_pct = roundHalfAway(pct, 1);
  }
  return out;
}

File: ios/CostSauceKit/Sources/CostSauceKit/Kernel/Kernel.swift
// The CostSauce costing kernel, Swift implementation — rounding, names,
// units, and money.
//
// Contract: docs/superpowers/specs/2026-07-25-native-ios-app-design.md
// §8-§10, pinned by shared/golden-vectors.json (also run by the Python and
// JavaScript kernels), compared as EXACT strings, zero tolerance. All
// decimals cross this API as STRINGS; internals are Int128 rationals
// (see Rational.swift). Rounding is half-away-from-zero everywhere.
//
// Function-for-function port of shared/kernel.js:35-128 (kernel.js is the
// reference implementation; line numbers below refer to it).

import Foundation

public enum Kernel {

    // MARK: - Rounding

    /// Round-half-away-from-zero at `places` decimals, formatted as a
    /// zero-padded decimal string. `q = floor((2n*10^places + d) / (2d))`
    /// on `|n|`; sign is re-applied unless `q == 0`. Port of kernel.js:35-45.
    public static func roundHalfAway(_ r: Rational, places: Int) -> String {
        let scale = Rational.pow10(places)
        var n = r.n
        let d = r.d
        let neg = n < 0
        if neg { n = -n }
        let q = (2 * n * scale + d) / (2 * d)
        let sign = (neg && q > 0) ? "-" : ""
        if places == 0 {
            return sign + String(q)
        }
        let ip = q / scale
        let fp = q % scale
        var fpString = String(fp)
        while fpString.count < places {
            fpString = "0" + fpString
        }
        return "\(sign)\(ip).\(fpString)"
    }

    // MARK: - Names

    /// lowercase → trim → drop non `[a-z0-9\s]` → collapse whitespace →
    /// strip a trailing "s" unless the word ends "ss" or is length ≤ 3.
    /// Port of kernel.js:48-54.
    public static func normalizeName(_ name: String) -> String {
        var s = name.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        s = s.replacing(/[^a-z0-9\s]/, with: "")
        s = s.replacing(/\s+/, with: " ")
        s = s.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.hasSuffix("s") && !s.hasSuffix("ss") && s.count > 3 {
            s.removeLast()
        }
        return s
    }

    public struct Candidate: Sendable {
        public let id: String
        public let name: String

        public init(id: String, name: String) {
            self.id = id
            self.name = name
        }
    }

    public enum MatchType: String {
        case exact, fuzzy
    }

    public struct Match {
        public let id: String
        public let name: String
        public let type: MatchType
    }

    /// Two passes in candidate order: exact normalized equality first, then
    /// bidirectional containment. Port of kernel.js:56-66.
    public static func matchIngredient(name: String, candidates: [Candidate]) -> Match? {
        let norm = normalizeName(name)
        if norm.isEmpty { return nil }
        for c in candidates where normalizeName(c.name) == norm {
            return Match(id: c.id, name: c.name, type: .exact)
        }
        for c in candidates {
            let cn = normalizeName(c.name)
            if cn.contains(norm) || norm.contains(cn) {
                return Match(id: c.id, name: c.name, type: .fuzzy)
            }
        }
        return nil
    }

    /// Mirrors api/routes/ingredients.py:63-65: candidates (already in
    /// (created_at, id) order) where the normalized query is a substring of
    /// the candidate's normalized name or vice versa, first 3; empty when
    /// the normalized query is empty.
    public static func nearMatches(name: String, candidates: [Candidate]) -> [Candidate] {
        let norm = normalizeName(name)
        if norm.isEmpty { return [] }
        var out: [Candidate] = []
        for c in candidates {
            let cn = normalizeName(c.name)
            if norm.contains(cn) || cn.contains(norm) {
                out.append(c)
                if out.count == 3 { break }
            }
        }
        return out
    }

    // MARK: - Units & money

    /// Exact-rational conversion factors to lb. Mirrors kernel.js:69-72's
    /// WEIGHT_TO_LB (computed there via parseDec at module load).
    private static let weightToLb: [String: Rational] = [
        "lb": Rational(n: 1, d: 1),
        "oz": Rational(n: 1, d: 16),
        "kg": Rational(n: 22_046_226_218, d: 10_000_000_000),
        "g": Rational(n: 22_046_226_218, d: 10_000_000_000_000),
    ]
    private static let baseUnits: Set<String> = ["lb", "oz", "kg", "g", "each"]

    /// qty entered in `unit` → quantity in the ingredient's `baseUnit`,
    /// rounded half-away-from-zero to 4dp. Exact port of kernel.js:75-102,
    /// including error message strings.
    public static func normalizePurchase(
        baseUnit: String, qty: String, unit: String,
        totalPrice: String, qtyInCase: String?
    ) throws -> String {
        let q = try Rational.parseDec(qty)
        let t = try Rational.parseDec(totalPrice)
        if !q.isPositive || !t.isPositive {
            throw KernelError("qty and total_price must be positive")
        }
        if !baseUnits.contains(baseUnit) {
            throw KernelError("unknown base_unit \(baseUnit)")
        }
        let normUnit = unit.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let baseQty: Rational
        if normUnit == "case" {
            guard let qtyInCase else {
                throw KernelError("qty_in_case required")
            }
            let qc = try Rational.parseDec(qtyInCase)
            if !qc.isPositive {
                throw KernelError("qty_in_case must be positive")
            }
            baseQty = q.mul(qc)
        } else if baseUnit == "each" {
            if normUnit != "each" {
                throw KernelError("tracked 'each' — use unit 'each' or 'case'")
            }
            baseQty = q
        } else {
            guard let unitRat = weightToLb[normUnit] else {
                throw KernelError("unsupported unit \(normUnit)")
            }
            guard let baseRat = weightToLb[baseUnit] else {
                throw KernelError("unknown base_unit \(baseUnit)")
            }
            baseQty = try q.mul(unitRat).div(baseRat)
        }
        let result = roundHalfAway(baseQty, places: 4)
        let resultRat = try Rational.parseDec(result)
        if resultRat.n <= 0 {
            throw KernelError("quantity is too small to register at 4 decimal places")
        }
        return result
    }

    /// Mirror of the DB generated column: round(total/qty, 6). Port of
    /// kernel.js:104-109.
    public static func unitPrice(totalPrice: String, qtyBaseUnits: String) throws -> String {
        let t = try Rational.parseDec(totalPrice)
        let q = try Rational.parseDec(qtyBaseUnits)
        if !t.isPositive || !q.isPositive {
            throw KernelError("total_price and qty_base_units must be positive")
        }
        return roundHalfAway(try t.div(q), places: 6)
    }

    /// `ceil(plate·10000 / (targetBp·50)) · 50` via integer math. Port of
    /// kernel.js:111-117.
    public static func suggestedPriceCents(plateCents: Int, targetBp: Int) throws -> Int {
        if targetBp <= 0 {
            throw KernelError("target_bp must be positive")
        }
        if plateCents < 0 {
            throw KernelError("plate_cents must be non-negative")
        }
        let num = Int128(plateCents) * 10000
        let den = Int128(targetBp) * 50
        let ceil = (num + den - 1) / den
        return Int(ceil * 50)
    }

    /// `fc = roundHalfAway(plate·100/menu, 1)`; status compares the ROUNDED
    /// value against `targetBp` (≤target "ok", ≤target+2 "watch", else
    /// "danger" — B4). Port of kernel.js:119-128.
    public static func fcStatus(
        plateCents: Int, menuCents: Int, targetBp: Int
    ) throws -> (fc: String, status: String) {
        if menuCents <= 0 {
            throw KernelError("menu_cents must be positive")
        }
        let plateRat = Rational(n: Int128(plateCents) * 100, d: Int128(menuCents))
        let fc = roundHalfAway(plateRat, places: 1)
        let fcR = try Rational.parseDec(fc)
        let target = Rational(n: Int128(targetBp), d: 100)
        let status: String
        if fcR.cmp(target) <= 0 {
            status = "ok"
        } else if fcR.cmp(target.add(Rational(n: 2, d: 1))) <= 0 {
            status = "watch"
        } else {
            status = "danger"
        }
        return (fc: fc, status: status)
    }
}
